shared/src/main/scala/pl/metastack/metarouter/Router.scala

```diff
@@ -69,16 +65,17 @@ object Router {
           }
       }
 
-    m[ROUTE](route.pathElements, split(s)).map { x =>
-      // Using asInstanceOf as a band aid since compiler isn't able to confirm the type.
-      RouteData(route, x.asInstanceOf[L])
-    }
+    m[ROUTE](route.pathElements, split(uri))
   }
 
-  def parse[ROUTE <: HList, T, Args <: HList](mappedRoute: MappedRoute[ROUTE, T], uri: String)
-                                             (implicit gen: Generic.Aux[T, Args],
-                                                       map: FlatMapper.Aux[Args.Convert.type, ROUTE, Args]): Option[T] =
-    parse(mappedRoute.route, uri).map(parsed => gen.from(parsed.data))
+  def parse[ROUTE <: HList, T, Args <: HList]
+    (mappedRoute: MappedRoute[ROUTE, T], uri: String)
+    (implicit gen: Generic.Aux[T, Args],
+              map: FlatMapper.Aux[Args.Convert.type, ROUTE, Args]): Option[T] =
+    parse(mappedRoute.route, uri).map { x =>
+      // Using asInstanceOf as a band aid since compiler isn't able to confirm the type.
+      gen.from(x.asInstanceOf[Args])
```

Comment:
`def m` should return `Option[X :< HList]`. You can know `X` on compile time by zipping the splitted URL with the route, filtering the route for `Arg`s and mapping the pairs, getting them parsed. HList [has](https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/syntax/hlists.scala) these methods.

The only difficulty will be to convert your `Seq[String]` with parsed url to `HList` since it is not known on compile time. I think this can be handled elegantly by casting and returning `Either[BadUrlError, Result]`.